### PR TITLE
Add sketch-file package

### LINKED ISSUE #168: Add fromFile / toFile functions
Create a `@sketch-hq/sketch-file` package, with `fromFile` and `toFile` methods we can use to read/parse Sketch documents, and save them.

### PR DESCRIPTION
**Pull request by bomberstudios**
Resolves #168

### CODE REVIEW

File: .changeset/blue-books-smash.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+"@sketch-hq/sketch-file": patch
```
Comment (christianklotz) [inline .changeset/blue-books-smash.md:2]: Curious to see how this is going to be versioned 😅 but I think with an initial release this could be almost anything `major`, `minor` or `patch`. As far as `sketch-file` is concerned it's "major" but the others are obviously unaffected.

In any case , changesets will preview the versioning later on. So all good 👍

**Review by christianklotz — COMMENTED** [review]
This is already looking quite nice, lets make sure to replace the `fromFile` in `validate-file.ts` with the new package (which should be added to `devDependencies`).

https://github.com/sketch-hq/sketch-document/blob/a0464fd5b29b29c22dede03a31219c56ef62a3fa/packages/file-format/scripts/validate-file.ts#L8-L38

File: packages/file-format/scripts/validate-file.ts [hunk 2]
```diff
@@ -3,43 +3,12 @@ import { ErrorObject } from 'ajv'
...
-      })
-    })
-  })
-
-  archive.close()
-
-  return contents
-}
+import schema from '../src/file-format.schema.json'
+import { fromFile, SketchFile } from '../../file'
```
Comment (christianklotz) [inline packages/file-format/scripts/validate-file.ts:7]: The beauty of the monorepo is that instead of relative imports you can treat `@sketch-hq/sketch-file` just like a normal dependency.

```suggestion
import { fromFile, SketchFile } from '@sketch-hq/sketch-file'
```

It would have to be added to the [`package.json`](https://github.com/sketch-hq/sketch-document/blob/50285d383ea4065a2b2f1df30eb0360823f3d914/packages/file-format/package.json) just like in the TypeScript types:
 
https://github.com/sketch-hq/sketch-document/blob/50285d383ea4065a2b2f1df30eb0360823f3d914/packages/file-format-ts/package.json#L23

Comment (bomberstudios) [reply packages/file-format/scripts/validate-file.ts:7]: I tried to do that but for some reason it would not work. Let me try again...

Comment (bomberstudios) [reply packages/file-format/scripts/validate-file.ts:7]: Ah, looks like I was missing the `package.json` when I tried, works now!

File: .changeset/blue-books-smash.md (hunk 1 above)
Comment (bomberstudios) [reply .changeset/blue-books-smash.md:2]: You’re right, looks like Changesets uses `patch` by default and that’s clearly not the case here. Will change.

File: packages/file/src/index.ts [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+      }
+      resolve(null)
+    })
+  })
+}
+
+/**
+ * Filter pages out of a SketchFile object based on page ids.
+ */
+const filterPages = (
```
Comment (jedrichards) [inline packages/file/src/index.ts:117]: This is neither used nor exported, maybe was copied in by mistake from Assistants?

File: packages/file/src/index.ts [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+import path from 'path'
+import StreamZip from 'node-stream-zip'
+import FileFormat from '../../file-format-ts'
+import Zip from 'adm-zip'
+
+/**
+ * Given a path to a Sketch file on the file system, this function unzips the
+ * JSON entries and parses them out into a SketchFile object.
+ */
+const fromFile = async (filepath: string): Promise<SketchFile> => {
```
Comment (jedrichards) [inline packages/file/src/index.ts:14]: Would it be nicer to have `fromFile` and `toFile` functions in their own files? Might be more scalable that way if new functions come in in future. No strong opinions either way at this stage though ...

File: packages/file/src/index.ts [hunk 5]
```diff
@@ -0,0 +1,142 @@
+/**
+ * This file is the entry point to {@link @sketch-hq/sketch-file}.
+ */
+
+import path from 'path'
+import StreamZip from 'node-stream-zip'
+import FileFormat from '../../file-format-ts'
```
Comment (jedrichards) [inline packages/file/src/index.ts:7]: Opportunity to import via the package name here, instead of relatively.

File: packages/file/src/index.ts [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+    )
+
+    // Write root level JSON data for document, user and meta data.
+    const data = {
+      document: JSON.stringify(<FileFormat.Document>{
+        ...contents.document,
+        pages: refs,
+      }),
+      user: JSON.stringify(contents.user),
+      meta: JSON.stringify(contents.meta),
```
Comment (jedrichards) [inline packages/file/src/index.ts:93]: Do we need to handle the optional `workspace` ref here too?

File: packages/file-format/package.json [hunk 7]
```diff
@@ -20,6 +20,7 @@
     "@types/json-schema": "7.0.7"
   },
   "devDependencies": {
+    "@sketch-hq/sketch-file-format": "6.0.1",
```
[the remark this answers is not in the document]
Comment (christianklotz) [reply packages/file-format/package.json:23]: This should be the following instead as far as I can tell:

```suggestion
    "@sketch-hq/sketch-file": "1.0.0",
```

(to cover the usage in `validate-file.ts`)

File: packages/file/README.md [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+
+Read and write Sketch documents from TypeScript.
+
+## Usage
+
+### `fromFile` function
+
+Read a Sketch Library file containing Color Variables, and dump them to JSON:
+
+```js
```
Comment (christianklotz) [inline packages/file/README.md:11]:
```suggestion
```ts
```

File: packages/file-format/package.json (hunk 7 above)
[the remark this answers is not in the document]
Comment (bomberstudios) [reply packages/file-format/package.json:23]: whoops

File: packages/file/src/index.ts (hunk 3 above)
Comment (bomberstudios) [reply packages/file/src/index.ts:117]: indeed, removing!

File: packages/file/README.md (hunk 8 above)
Comment (bomberstudios) [reply packages/file/README.md:11]: 🤙🏻

File: packages/file/jest.config.js [hunk 9]
```diff
@@ -0,0 +1,4 @@
+module.exports = {
+  preset: 'ts-jest',
+  testEnvironment: 'node',
```
Comment (christianklotz) [inline packages/file/jest.config.js:3]: Might be nice to be consistent and use the exact same config as in https://github.com/sketch-hq/sketch-document/blob/a0464fd5b29b29c22dede03a31219c56ef62a3fa/packages/file-format-ts/jest.config.js#L1-L5

```suggestion
  testEnvironment: 'node',
  testRegex: '^.*\\.test.ts$',
```

File: packages/file/src/index.ts (hunk 6 above)
Comment (bomberstudios) [reply packages/file/src/index.ts:93]: Good point, let’s discuss this and the `fromFile` / `toFile` separation in Slack

File: packages/file/src/file.ts [hunk 10]
```diff
@@ -0,0 +1,125 @@
+import path from 'path'
+import StreamZip from 'node-stream-zip'
+import FileFormat from '@sketch-hq/sketch-file-format-ts'
+import Zip from 'adm-zip'
+
+/**
+ * Given a path to a Sketch file on the file system, this function unzips the
+ * JSON entries and parses them out into a SketchFile object.
+ */
```
Comment (christianklotz) [inline packages/file/src/file.ts:9]: Maybe good opportunity to revisit the comment to provide better documentation:

```suggestion
/**
 * Returns a {@link Promise} that resolves with a fully parsed {@link SketchFile}
 * object once the file at the given path has been unzipped and read successfully.
 *
 * @param filepath -  The file path to the Sketch document
 */
```

File: packages/file/src/file.ts [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+
+  archive.close()
+
+  return { filepath, contents }
+}
+
+/**
+ * Given a SketchFile object, this function saves a valid Sketch document.
+ */
+const toFile = async (obj: SketchFile): Promise<void> => {
```
Comment (christianklotz) [inline packages/file/src/file.ts:56]: Similarly here, [TSDoc](https://tsdoc.org/play) could do with a little more detail here.

**Review by christianklotz — COMMENTED** [review]
Nice, just a few comments around documentation/comments – especially important since we want this to be as accessible to others as possible.

File: packages/file/README.md [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+          r: color.value.red,
+          g: color.value.green,
+          b: color.value.blue,
+          a: color.value.alpha,
+        },
+      })
+    })
+    console.log(JSON.stringify(colors, null, 2))
+  }
+})
```
Comment (christianklotz) [inline packages/file/README.md:43]: I'd suggest to move the custom type out of the transformation and add a couple of comments to make it clear what the sample code represents. And bail early for leftmost indentation.

```suggestion
// Custom type color swatches should be converted to
type ColorToken = {
  name: string
  value: {
    r: number
    g: number
    b: number
    a: number
  }
}

// Transform swatches into list of `ColorToken` after parsing the file successfully
fromFile(resolve(__dirname, sketchDocumentPath)).then((parsedFile) => {
  const doc = parsedFile.contents.document
  if (!doc.sharedSwatches) return
  
  var colors: Array<ColorToken> = []
  doc.sharedSwatches.objects.forEach((color) => {
    colors.push({
      name: color.name,
      value: {
        r: color.value.red,
        g: color.value.green,
        b: color.value.blue,
        a: color.value.alpha,
      },
    })
  })
  
  console.log(JSON.stringify(colors, null, 2))
})
```

**Review by christianklotz — APPROVED** [review]
Implementation looking good. Approving on the condition to amend the README slightly

**Comment by bomberstudios** [comment]
🎉
